custom_metrics/javascript.js

```diff
@@ -0,0 +1,73 @@
+//[javascript]
+// Uncomment the previous line for testing on webpagetest.org
+
+// Please, refer to instructions for adding a custom metric in almanac.js.
+
+const requests = $WPT_REQUESTS;
+
+return JSON.stringify({
+    ajax_requests: (() => {
+        // Returns the number of ajax requests per page.
+        var ajax_apis = ['xmlhttprequest', 'fetch', 'beacon'];
+        return window.performance.getEntriesByType('resource').filter(r => ajax_apis.includes(r.initiatorType)).reduce((obj, r) => {
+            obj.total++;
+            obj[r.initiatorType]++;
+            return obj;
+        }, Object.fromEntries([...ajax_apis, 'total'].map(api => [api, 0])))
+    })(),
+
+    // Returns the number of ajax requests using beacon.
+    beacon_ajax_usage: window.performance.getEntriesByType('resource').filter((r) => r.initiatorType === 'beacon').length,
+
+    // Returns the number of iframes per page.
+    iframe: document.getElementsByTagName('iframe').length,
+
+    requests_protocol: (() => {
+        // Returns the percentage of http 2 protocol used.
+        try {
+            var ajaxs = window.performance.getEntriesByType('resource').filter((r) => ['xmlhttprequest', 'fetch', 'beacon'].includes(r.initiatorType));
+            var resources = window.performance.getEntriesByType('resource').filter((r) => r.initiatorType !== 'xmlhttprequest' && r.initiatorType !== 'fetch' && r.initiatorType !== 'beacon');
+            let request_data = {
+                ajax_h1: ajaxs.filter(r => ['http/1.1', 'http/1'].includes(r.nextHopProtocol)).length,
+                resources_h1: resources.filter(r => ['http/1.1', 'http/1'].includes(r.nextHopProtocol)).length,
+                ajax_h2: ajaxs.filter(r => ['h2', 'http/2', 'http2'].includes(r.nextHopProtocol)).length,
+                resources_h2: resources.filter(r => ['h2', 'http/2', 'http2'].includes(r.nextHopProtocol)).length,
+                ajax_h3: ajaxs.filter(r => ["h3", "http/3", "h3-29"].includes(r.nextHopProtocol)).length,
+                resources_h3: resources.filter(r => ["h3", "http/3", "h3-29"].includes(r.nextHopProtocol)).length,
+            }
+            return request_data;
+        } catch (e) {
+            return null;
+        }
+    })(),
+    // Returns the content-types of requests.
+    content_types: requests.map((req) => req.response_headers['content-type']),
+
+    script_tags: (() => {
+        let script_tags = Array.from(document.querySelectorAll('script'));
+
+        let script_data = {
+            total: script_tags.length,
+            async: script_tags.filter(tag => tag.async).length,
+            defer: script_tags.filter(tag => tag.defer).length,
```

Comment:
One thing the 2020 chapter looked at was the usage of both [async and defer](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2020/javascript/async_and_defer.sql) on the same script. Would be good to support that in the custom metric.

Also given the [special note](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-defer) for `defer` on MDN about the attribute not being used if the `src` attribute is absent, it may be nice to include a metric for scripts that use that antipattern.

```suggestion
            async: script_tags.filter(tag => tag.async).length,
            defer: script_tags.filter(tag => tag.defer).length,
            async_and_defer: script_tags.filter(tag => tag.async && tag.defer).length,
            defer_without_src: script_tags.filter(tag => tag.defer && !tag.src).length,
```

Reply:
Isn't this already covered as a metric due to the 2020 chapter? Do we need to rewrite this? @rviscomi

Reply:
added these anyway

Reply:
It was covered last year but the problem was that the custom metric didn't exist and we (I was the analyst) implemented it by parsing the HTML in SQL, which is expensive and error-prone. It'd be much better to process the data in JS instead.